packages/ember-glimmer/tests/integration/helpers/readonly-test.js

```diff
@@ -0,0 +1,163 @@
+import { RenderingTest, moduleFor } from '../../utils/test-case';
+import { Component } from '../../utils/helpers';
+import { set } from 'ember-metal/property_set';
+import { get } from 'ember-metal/property_get';
+
+moduleFor('Helpers test: {{readonly}}', class extends RenderingTest {
+
+  ['@test {{readonly}} of a path should work']() {
+    let component;
+
+    this.registerComponent('foo-bar', {
+      ComponentClass: Component.extend({
+        didInsertElement() {
+          component = this;
+        }
+      }),
+      template: '{{value}}'
+    });
+
+    this.render('{{foo-bar value=(readonly val)}}', {
+      val: 12
+    });
+
+    this.assertText('12');
+
+    this.assertStableRerender();
+
+    if (this.isGlimmer) {
+      expectAssertion(() => set(component, 'value', 13), /cannot update the value of a readonly/);
+    } else {
+      this.assert.notOk(component.attrs.value.update);
+    }
+
+    this.assertText('12', 'local property is not updated');
+    this.assert.equal(get(this.context, 'val'), 12, 'upstream attribute is not updated');
+  }
+
+  ['@test {{readonly}} of a string renders correctly']() {
+    let component;
+
+    this.registerComponent('foo-bar', {
+      ComponentClass: Component.extend({
+        didInsertElement() {
+          component = this;
+        }
+      }),
+      template: '{{value}}'
+    });
+
+    this.render('{{foo-bar value=(readonly "12")}}');
+
+    this.assertText('12');
+
+    this.assertStableRerender();
+
+    if (this.isGlimmer) {
+      expectAssertion(() => set(component, 'value', 13), /cannot update the value of a readonly/);
+    } else {
+      this.assert.notOk(component.attrs.value.update);
+    }
+
+    this.assertText('12', 'local property is not updated');
+  }
+
+  ['@test {{readonly}} of a {{mut}} does not return a mutable cell']() {
+    let component;
+
+    this.registerComponent('foo-bar', {
+      ComponentClass: Component.extend({
+        didInsertElement() {
+          component = this;
+        }
+      }),
+      template: '{{value}}'
+    });
+
+    this.render('{{foo-bar value=(readonly (mut val))}}', {
+      val: 12
+    });
+
+    this.assertText('12');
+
+    this.assertStableRerender();
+
+    if (this.isGlimmer) {
+      expectAssertion(() => set(component, 'value', 13), /cannot update the value of a readonly/);
+    } else {
+      this.assert.notOk(component.attrs.value.update);
+    }
+
+    this.assertText('12', 'local property is not updated');
+    this.assert.equal(get(this.context, 'val'), 12, 'upstream attribute is not updated');
+  }
+
+  ['@htmlbars {{mut}} of a {{readonly}} mutates only the middle and bottom tiers']() {
```

Comment:
@krisselden @chancancode so this is still failing in glimmer.

To make it backwards compatible for curly components, what ember-htmlbars seems to be doing is disconnecting the readonly if it is updated from a child, such that `get(child, 'parentAttr')` is no longer the same as `child.attr.parentAttr`. But when a change comes from the upstream property, the value correctly updates.

I've tried a variety of ways of disconnecting the value from the reference and reconnecting it, but nothing I've tried really works .It also seems like there could be multiple places this logic can go into (should it be in `process-args` or `mut`?) Any direction here would be helpful.